### ISSUE TITLE
딥러닝 모델링

### ISSUE THREAD

**Issue by Jangchanhi** [Jangchanhi/pyAICE#5]
딥러닝 학습 방법 : 딥러닝 모델의 매개 변수(weight, bias)를 무작위로 부여한 후, 반복학습을 통해 모델의 출력값을 정답과 일치하도록 매개변수를 조금씩 조정함.
	-> 최적화 알고리즘(Gradient Descent)

perceptron : 사람 두뇌에 있는 뉴런을 모델링 한것 : 간단한 함수를 학습할 수 있음.

DNN (Deep Neural Network) : 입력층과 출력층 사이에 여러 개의 은닉층(hidden layer)으로 이루어진 인공신경망 -> 신경망 출력에 비선형 활성화 함수를 추가하여 복잡한 비선형 관계를 모델링 할 수 있음.

loss Function : 신경망 학습의 목적 함수로 출력값(예측값)과 정담(실제값)의 차이를 계산
1) 회귀 모델 시 : MSE, MAE 값이 있음
2) 분류 모델 시 : 이진분류(Binary cross-entropy), 다중 분류(Categorical cross-entropy)가 있음

Gradient Descent : 뉴럴넷이 가중치 파라미터들을 최적화하는 방법
-> 손실함수의 현 가중치에서 기울기를 구해서 Loss를 줄이는 방향으로 업데이트 해 나감

Back Propagation : 실제값과 모델 결과값에서 오차를 구해서 오차를 output에서 input 방향으로 보냄 -> 가중치를 재업데이트 하면서 학습

Optimization Algorithm

[딥러닝 주요 알고리즘]
<CNN> : 가중치 필터로 학습되며, 이미지 특징/패턴 찾아내 이미지 분류가 가능함
1) DNN : 이미지 처리를 DNN(fully connected neual network) 수행 시 문제점
- 이미지 1차원 형태로 처리 : 위치정보 소실 -> 정보소실
- 이미지 사이즈가 커질 경우 비례해서 학습할 가중치 증가

2) CNN : (convolutional Neural Network) : 이미지의 특징을 추출하는 부분과 클래스를 분류하는 부분
- 특징 추출 영역 : Convolution layer와 Pooling layer를 여러 겹 쌓는 형태로 구성
- 클래스 분류 영역 : Fully Connected layer 구성하여 사용

[ padding ] : convolutional layer의 출력 데이터가 줄어드는 것을 방지하는 방법
 -> 입력 데이터의 외곽에 0으로 채워 넣는 것

[Pooling layer]

[RNN] : 입력과 출력을 시퀸스 단위로 처리하는 모델 
 -> 음성인식, 언어 모델링, 번역 ,이미지 주석 생성에 활용
-히든 노드가 방향을 가진 엣지로 연결돼 순환구조(directed cycle)를 이루는 인공신경망의 한 종류
<RNN의 다양한 구조> one to one, one to many(사진설명 붙이기), many to one(감성분류), many to many(번역)

<RNN의 문제점(vanishing gradient problem)>
1) LSTM : 오랫동안 정보를 전달할 수 있음
-> 문장과 같은 단어가 문장 안에서 순서가 중요한 경우나 주가와 같은 시계열 데이터 효과
- LSTM(Long Short-Term Memory) : RNN은 관련 정보와 그 정보를 사용하는 지점 사이 거리가 멀 경우 역전파시 그래디언트가 점차 줄어 학습능력이 크게 저하되는 것으로 알려져 있다.
- Rnn의 히든 state에 cell-state를 추가한 구조
